### PR TITLE
[SPARK-20371][R] Add wrappers for collect_list and collect_set

### PR DESCRIPTION
**Pull request by zero323**
## What changes were proposed in this pull request?

Adds wrappers for `collect_list` and `collect_set`.

## How was this patch tested?

Unit tests, `check-cran.sh`

### CODE REVIEW

**Comment by zero323** [comment]
cc @felixcheung

File: R/pkg/R/functions.R [hunk 1]
```diff
@@ -3652,3 +3652,43 @@ setMethod("posexplode",
             jc <- callJStatic("org.apache.spark.sql.functions", "posexplode", x@jc)
             column(jc)
           })
+
+#' collect_list
+#'
+#' Aggregate function: returns a list of objects with duplicates.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: does other function has `Aggregate function: ` in the description or this is carried over from Scala doc?
if latter we could go without - there's already a `@family` tag

File: R/pkg/R/functions.R [hunk 2]
```diff
@@ -3652,3 +3652,43 @@ setMethod("posexplode",
...
+#'
+#' @param x Column to compute on
+#'
+#' @rdname collect_list
+#' @name collect_list
+#' @family aggregate_functions
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: I think there's an existing `agg_funcs` in R

File: R/pkg/R/generics.R [hunk 3]
```diff
@@ -918,6 +918,14 @@ setGeneric("cbrt", function(x) { standardGeneric("cbrt") })
 #' @export
 setGeneric("ceil", function(x) { standardGeneric("ceil") })
 
+#' @rdname collect_list
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: these are under `###################### Expression Function Methods ##########################`
which doesn't seem like the right group

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 4]
```diff
@@ -1714,6 +1714,28 @@ test_that("group by, agg functions", {
   expect_true(abs(sd(1:2) - 0.7071068) < 1e-6)
   expect_true(abs(var(1:5, 1:5) - 2.5) < 1e-6)
 
+  # Test collect_list and collect_set
+  gd3_collections_local <- collect(
+    agg(gd3, collect_set(df8$age), collect_list(df8$age))
+  )
+
+  testthat::expect_equal(
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: why `testthat::`?

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/R/functions.R (hunk 1 above)
Comment (zero323) [inline R/pkg/R/functions.R]: From Scala docs. Removed.

File: R/pkg/R/functions.R (hunk 2 above)
Comment (zero323) [inline R/pkg/R/functions.R]: Yes, corrected.

File: R/pkg/R/generics.R (hunk 3 above)
Comment (zero323) [inline R/pkg/R/generics.R]: Let's continue here https://github.com/apache/spark/pull/17674#discussion_r112172948

File: R/pkg/inst/tests/testthat/test_sparkSQL.R (hunk 4 above)
Comment (zero323) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Just a habit. Corrected.

File: R/pkg/R/functions.R (hunk 2 above)
Comment (felixcheung) [inline R/pkg/R/functions.R]: re: my other comments, my preference would be full text without underscore like `@family aggregate functions`. This isn't a key or id and it shows up in generated doc text. Anyway, that's a bigger change all around.

File: R/pkg/R/generics.R (hunk 3 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: ok, it's good.

File: R/pkg/R/functions.R [hunk 5]
```diff
@@ -3652,3 +3652,43 @@ setMethod("posexplode",
...
+
+#' collect_list
+#'
+#' Creates a list of objects with duplicates.
+#'
+#' @param x Column to compute on
+#'
+#' @rdname collect_list
+#' @name collect_list
+#' @family agg_func
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: it has an 's' though `agg_funcs` - not to nit pick but it needs to match exactly

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/R/functions.R (hunk 5 above)
Comment (zero323) [inline R/pkg/R/functions.R]: My bad.

On a side note, do you think we should provide detailed examples for each function?

Comment (felixcheung) [inline R/pkg/R/functions.R]: we really should - it's a long standing concern that something like 
```
@examples \dontrun{collect_list(df$x)}
```
is not useful at all. At least we should include how this goes into a select statement, and in some cases what the output looks like. However, with the number of sql functions we have it's a bit of ongoing work to improve this and we definitely can use more help on that

**Comment by felixcheung** [comment]
btw, fyi, you don't have to rebase or squash each time you push - it's actually easier to review if you don't - so reviewer can track comments and see the diff from the last time.

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by zero323** [comment]
> btw, fyi, you don't have to rebase or squash each time you push

Noted :) But yeah, there was a conflict after SPARK-20375.

>  At least we should include how this goes into a select statement, and in some cases what the output looks like. 

It would be great.

**Comment by felixcheung** [comment]
merged to master.

**Comment by zero323** [comment]
Thanks @felixcheung

**Comment by zero323** [comment]
BTW @felixcheung - is there any deeper reason behind current stat of `generics.R`? I mean:

- Inconsistent usage of standard and `roxygen` comments.
- Marking functions which are not to be exported with `@export`.
- Slightly mixed up order (both in groups and between groups).
- Some minor inconsistencies (like marking `asc` as `@rdname columnfunctions`).

**Comment by felixcheung** [comment]
Not really its just inconsistent handling.
Some comment changes can be deliberated though.

**Comment by zero323** [comment]
Yeah, I have this feeling that it could be deliberate, but I cannot figure out what is the purpose. Removing `@exports` should be enough, shouldn't it?

I thought about cleaning this up, but I wonder if it is better to wait for [SPARK-16693](https://issues.apache.org/jira/browse/SPARK-16693).

**Comment by felixcheung** [comment]
Will be probably cleaner

**Comment by zero323** [comment]
@felixcheung Do you know by any chance what is the policy about adding new datasets to Spark? License restrictions, file size and such?

**Comment by felixcheung** [comment]
@zero323 I think that its license needs to be compatible with Apache 2.0 and it can't be big (since example data is in the release; no more than a few MB?) https://www.apache.org/licenses/

**Pull request closed without merging**